signing/signing.go

```diff
@@ -1,20 +1,387 @@
+			repoPath = string(ext.Value)
+		}
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.3" {
+			repoSHA = string(ext.Value)
+		}
+	}
+
+	if len(repoRef) == 0 || len(repoPath) == 0 {
+		return "", "", "", "", fmt.Errorf("error extracting repo ref or path from certificate %v", err)
+	}
+
+	// Get workflow job ref from the certificate.
+	certURIs := cert.URIs
+	if len(certURIs) == 0 {
+		return "", "", "", "", errors.New("error: certificate has no URIs")
+	}
+	workflowRef := certURIs[0].Path
+	if len(workflowRef) == 0 {
+		return "", "", "", "", errors.New("error: workflow path is empty")
+	}
+
+	// Remove repo path from workflow filepath.
+	ind := strings.Index(workflowRef, repoPath) + len(repoPath) + 1
+	workflowPath = workflowRef[ind:]
+
+	// Remove repo ref tag from workflow filepath.
+	workflowPath = workflowPath[:strings.Index(workflowPath, "@")]
+
+	return repoPath, repoRef, repoSHA, workflowPath, nil
+}
+
+func verifyScorecardWorkflow(workflowContent string) error {
+	// Verify workflow contents using actionlint.
+	workflow, lintErrs := actionlint.Parse([]byte(workflowContent))
+	if lintErrs != nil {
+		return fmt.Errorf("actionlint errors parsing workflow: %v", lintErrs)
+	}
+
+	// Verify that there are no global env vars or defaults.
+	if workflow.Env != nil || workflow.Defaults != nil {
+		return errors.New("workflow contains global env vars or defaults")
+	}
+
+	// Verify that the only global permission set is read-all.
+	if workflow.Permissions.All.Value != "read" && workflow.Permissions.All.Value != "read-all" {
+		return errors.New("workflow global permission isn't read-all")
+	}
+
+	// Find the job with a step that calls scorecard-action.
+	var scorecardJob *actionlint.Job
+	for _, job := range workflow.Jobs {
+		scorecardStepFound := false
+		if !scorecardStepFound {
+			for _, step := range job.Steps {
+				stepName := step.Exec.(*actionlint.ExecAction).Uses.Value
+				stepName = stepName[:strings.Index(stepName, "@")]                                    // get rid of commit sha.
+				if stepName == "ossf/scorecard-action" || stepName == "rohankh532/scorecard-action" { // TODO: remove rohankh532 later.
+					scorecardJob = job
+					scorecardStepFound = true
+					break
+				}
+			}
+		}
+		// Make sure other jobs don't have id-token permissions.
+		if !scorecardStepFound && job.Permissions != nil {
+			idToken := job.Permissions.Scopes["id-token"]
+			if idToken != nil && idToken.Value.Value == "write" {
+				return errors.New("workflow has a non-scorecard job with id-token permissions")
+			}
+		}
+	}
+
+	if scorecardJob == nil {
+		return errors.New("workflow has no job calling ossf/scorecard-action")
+	}
+
+	// Verify that there is no job container or services.
+	if scorecardJob.Container != nil || len(scorecardJob.Services) > 0 {
+		return errors.New("workflow contains container or service")
+	}
+
+	// Verify that the workflow runs on ubuntu-latest and nothing else.
+	if scorecardJob.RunsOn == nil {
```

Comment:
is it necessary to have a `runs-on`? If it is, can you add a comment to the code? Otherwise, don't return error

Reply:
I get a syntax error when deleting the runs-on from the workflow file, so I think its necessary to define ubuntu.

Reply:
Thanks. Can you please add a comment to the code?